doc/guide/transactions-and-threading.rst

```diff
@@ -0,0 +1,394 @@
+   >>> db.history(conn.root()._p_oid)[0]['description']
+   'incrementing x again'
+
+Here, we used the
+:meth:`~transaction.interfaces.ITransactionManager.get` method to get
+the current transaction.
+
+Connection isolation
+--------------------
+
+In the last few examples, we used a connection opened using
+:meth:`~ZODB.DB.transaction`.  This was distinct from and used a
+different transaction manager than the original connection. If we
+looked at the original connection, ``conn``, we'd see that it has the
+same value for ``x`` that we set earlier:
+
+  >>> conn.root.x
+  3
+
+This is because it's still in the same transaction that was implicitly
+begun when a change was last committed against it.  If we want to see
+changes, we have to begin a new transaction:
+
+  >>> trans = tm.begin()
+  >>> conn.root.x
+  5
+
+ZODB uses a timestamp-based commit protocol that provides `snapshot
+isolation <https://en.wikipedia.org/wiki/Snapshot_isolation>`_.
+Whenever we look at ZODB data, we see its state as of the time the
+transaction began.
+
+.. _conflicts-label:
+
+Conflict errors
+---------------
+
+As mentioned in the previous section, each connection sees and
+operates on a view of the database as of the transaction start time.
+If two connections modify the same object at the same time, one of the
+connections will get a conflict error when it tries to commit::
+
+  with db.transaction() as conn2:
+     conn2.root.x += 1
+
+  conn.root.x = 9
+  tm.commit() # will raise a conflict error
+
+.. -> src
+
+    >>> exec(src) # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+    ...
+    ZODB.POSException.ConflictError: ...
+
+If we executed this code, we'd get ``ConflictError`` exception on the
+last line.  After a conflict error is raised, we'd need to abort the
+transaction, or begin a new one, at which point we'd see the data as
+written by the other connection:
+
+    >>> tm.abort()
+    >>> conn.root.x
+    6
+
+The timestamp-based approach used by ZODB is referred to as an
+*optimistic* approach, because it works best if there are no
+conflicts.
+
+The best way to avoid conflicts is to design your application so that
+multiple connections don't update the same object at the same time.
+This isn't always easy.
+
+Sometimes you may need to queue some operations that update shared
+data structures, like indexes, so the updates can be made by a
+dedicated thread or process.
+
+Conflict resolution
+~~~~~~~~~~~~~~~~~~~
+
+ZODB provides a conflict-resolution framework for merging conflicting
+changes. This is implemented by `BTree
+<https://pythonhosted.org/BTrees/>`_ buckets and ``Length`` objects.
+
+The main data structures provided by BTrees: BTrees and TreeSets,
+spread their data over multiple objects.  The leaf-level objects,
+called *buckets* allow distinct keys to be updated without causing
+conflicts [#usually-avoids-conflicts]_.
+
+``Length`` objects are conflict key counters, that merge changes by
+simply accumulating changes.
```

Comment:
Shout-out to https://github.com/zopefoundation/zope.minmax ?
zc.queue also resolves conflicts, maybe that deserves a pointer here and/or in the section above about queueing tasks. (zc.catalogqueue does too, but that's probably too specialized to mention...well, maybe not, you did mention indexing.)